### PR TITLE
chore(api,worker): chat oauth flow allow to provide the integration identifier

### PR DESCRIPTION
**Pull request by LetItRock**
### What change does this PR introduce?

We are going to allow the creation of multiple integrations for the same channel and provider.
This PR touches the chat OAuth code and logic that updates the subscriber credentials after the successful OAuth flow.

Allow passing the integration identifier that will be used when we store the credentials for the integration:
- in the chat OAuth flow
- in the update subscriber credentials endpoint

### Why was this change needed?

Multi Channel Providers PRD.

### CODE REVIEW

File: apps/api/src/app/subscribers/dtos/chat-oauth-request.dto.ts [hunk 1]
```diff
@@ -7,6 +7,10 @@ export class ChatOauthRequestDto {
   @IsMongoId()
   @IsDefined()
   environmentId: string;
+
+  @IsOptional()
+  @IsString()
+  integrationIdentifier?: string;
```
Comment (LetItRock) [inline apps/api/src/app/subscribers/dtos/chat-oauth-request.dto.ts:13]: The chat OAuth flow - allow providing the `integration identifier` that will be used to find the exact integration, then the chat channel credentials are stored under the subscriber credentials array for that specific integration.

If not provided then we will be looking for the latest created active integration, which is backward compatible change.

File: apps/api/src/app/subscribers/dtos/update-subscriber-channel-request.dto.ts [hunk 2]
```diff
@@ -12,6 +13,14 @@ export class UpdateSubscriberChannelRequestDto {
   @IsDefined()
   providerId: ChatProviderIdEnum | PushProviderIdEnum;
 
+  @ApiProperty({
+    type: String,
+    description: 'The integration identifier',
+  })
+  @IsString()
+  @IsOptional()
+  integrationIdentifier?: string;
```
Comment (LetItRock) [inline apps/api/src/app/subscribers/dtos/update-subscriber-channel-request.dto.ts:22]: also the update subscriber credentials endpoint allows the same

File: apps/api/src/app/subscribers/usecases/chat-oauth-callback/chat-oauth-callback.usecase.ts [hunk 3]
```diff
@@ -136,7 +141,13 @@ export class ChatOauthCallback {
...
     };
 
-    const integration = await this.integrationRepository.findOne(query);
+    if (command.integrationIdentifier) {
+      query.identifier = command.integrationIdentifier;
+    }
+
+    const integration = await this.integrationRepository.findOne(query, undefined, {
+      query: { sort: { createdAt: -1 } },
+    });
```
Comment (LetItRock) [inline apps/api/src/app/subscribers/usecases/chat-oauth-callback/chat-oauth-callback.usecase.ts:150]: find the latest created integration or by identifier if specified

File: apps/api/src/app/subscribers/usecases/chat-oauth-callback/chat-oauth-callback.usecase.ts [hunk 4]
```diff
@@ -96,10 +97,14 @@ export class ChatOauthCallback {
     command: ChatOauthCallbackCommand,
     integrationCredentials: ICredentialsDto
   ): Promise<string> {
-    const redirectUri =
+    let redirectUri =
       process.env.API_ROOT_URL +
       `/v1/subscribers/${command.subscriberId}/credentials/${command.providerId}/oauth/callback?environmentId=${command.environmentId}`;
 
+    if (command.integrationIdentifier) {
+      redirectUri = `${redirectUri}&integrationIdentifier=${command.integrationIdentifier}`;
```
Comment (LetItRock) [inline apps/api/src/app/subscribers/usecases/chat-oauth-callback/chat-oauth-callback.usecase.ts:105]: pass `integrationIdentifier` to the `/oauth/callback` endpoint

File: apps/api/src/app/subscribers/usecases/chat-oauth/chat-oauth.usecase.ts [hunk 5]
```diff
@@ -54,10 +54,23 @@ export class ChatOauth {
...
+      `/v1/subscribers/${subscriberId}/credentials/slack/oauth/callback?environmentId=${environmentId}`;
+
+    if (integrationIdentifier) {
+      redirectUri = `${redirectUri}&integrationIdentifier=${integrationIdentifier}`;
+    }
 
-    return `${this.SLACK_OAUTH_URL}client_id=${clientId}&scope=incoming-webhook&user_scope=&redirect_uri=${redirectUri}`;
+    return `${
+      this.SLACK_OAUTH_URL
+    }client_id=${clientId}&scope=incoming-webhook&user_scope=&redirect_uri=${encodeURIComponent(redirectUri)}`;
```
Comment (LetItRock) [inline apps/api/src/app/subscribers/usecases/chat-oauth/chat-oauth.usecase.ts:73]: the `redirect_uri` should be URI encoded

File: apps/api/src/app/subscribers/usecases/update-subscriber-channel/update-subscriber-channel.spec.ts [hunk 6]
```diff
@@ -121,4 +123,40 @@ describe('Update Subscriber channel credentials', function () {
     expect(newChannel?.providerId).to.equal('slack');
     expect(newChannel?.credentials.webhookUrl).to.equal('new-secret-webhookUrl');
   });
+
+  it('should update slack channel credentials for a specific integration', async function () {
```
Comment (LetItRock) [inline apps/api/src/app/subscribers/usecases/update-subscriber-channel/update-subscriber-channel.spec.ts:127]: added the test to cover the new logic on the use-case

File: apps/api/src/app/subscribers/usecases/update-subscriber-channel/update-subscriber-channel.usecase.ts [hunk 7]
```diff
@@ -39,7 +52,8 @@ export class UpdateSubscriberChannel {
     const updatePayload = this.createUpdatePayload(command);
 
     const existingChannel = foundSubscriber?.channels?.find(
-      (subscriberChannel) => subscriberChannel.providerId === command.providerId
+      (subscriberChannel) =>
+        subscriberChannel.providerId === command.providerId && subscriberChannel._integrationId === foundIntegration._id
```
Comment (LetItRock) [inline apps/api/src/app/subscribers/usecases/update-subscriber-channel/update-subscriber-channel.usecase.ts:56]: update the subscriber credentials for the specific integration

File: apps/worker/src/app/workflow/usecases/send-message/send-message-chat.usecase.ts [hunk 8]
```diff
@@ -137,10 +137,11 @@ export class SendMessageChat extends SendMessageBase {
   private async sendChannelMessage(
     command: SendMessageCommand,
     subscriberChannel: IChannelSettings,
-    chatChannel,
+    chatChannel: NotificationStepEntity,
     content: string
   ) {
     const integration = await this.getIntegration({
+      id: subscriberChannel._integrationId,
```
Comment (LetItRock) [inline apps/worker/src/app/workflow/usecases/send-message/send-message-chat.usecase.ts:144]: When we do send the chat message we are iterating over the subscriber channels and for each channel we do send the message. Here we try to find the integration by the `integrationId` from the subscriber channels array. The subscriber channel array stores the webhook URLs where the message is sent.

File: docs/docs/channels/chat/slack.md [hunk 9]
```diff
@@ -23,13 +23,36 @@ This step is optional, if you already have a Slack application you can reuse it.
...
+   The `Shareable URL` should have the following format:
+
+   ```bash
+   https://api.novu.co/v1/subscribers/SUBSCRIBER_ID/credentials/slack/oauth?environmentId=ENVIRONMENT_ID&integrationIdentifier=INTEGRATION_IDENTIFIER.
+   ```
+
+   - SUBSCRIBER_ID is a custom identifier used when identifying your users within the Novu platform. [Read more here](/platform/subscribers). <br/>
+   - ENVIRONMENT_ID is a context of an environment you can locate your environment id in the setting in the dashboard . <a href="https://web.novu.co/settings" style={{textDecoration: "underline"}}>Settings</a>.
+     <br/>
+   - INTEGRATION_IDENTIFIER optional, is a unique identifier of the integration. You can locate your integration identifier in the <a href="https://web.novu.co/integrations" style={{textDecoration: "underline"}}>Integration Store</a>. When not provided the last created integration will be used.
```
Comment (djabarovgeorge) [inline docs/docs/channels/chat/slack.md:38]:
```
When not provided the last created integration will be used.
```


Should we inform about the side effects that may arise from this?

File: apps/api/src/app/subscribers/dtos/chat-oauth-request.dto.ts [hunk 10]
```diff
@@ -7,6 +7,10 @@ export class ChatOauthRequestDto {
   @IsMongoId()
   @IsDefined()
   environmentId: string;
+
+  @IsOptional()
```
Comment (djabarovgeorge) [inline apps/api/src/app/subscribers/dtos/chat-oauth-request.dto.ts:11]: this one is optional in the flow so we won't make breaking changes, right?
if this is the case i think we still need to break it in a couple of versions.

File: apps/api/src/app/subscribers/usecases/chat-oauth-callback/chat-oauth-callback.usecase.ts [hunk 11]
```diff
@@ -96,10 +97,14 @@ export class ChatOauthCallback {
     command: ChatOauthCallbackCommand,
     integrationCredentials: ICredentialsDto
   ): Promise<string> {
-    const redirectUri =
+    let redirectUri =
       process.env.API_ROOT_URL +
       `/v1/subscribers/${command.subscriberId}/credentials/${command.providerId}/oauth/callback?environmentId=${command.environmentId}`;
```
Comment (djabarovgeorge) [inline apps/api/src/app/subscribers/usecases/chat-oauth-callback/chat-oauth-callback.usecase.ts:102]: The only thing is maybe we could extract the builder of redirectUri to some shared file so both usecases could use it, my mistake not doing that but it is not a must at all.

Comment (djabarovgeorge) [reply apps/api/src/app/subscribers/usecases/chat-oauth-callback/chat-oauth-callback.usecase.ts:102]: Maybe we could extract the builder of redirectUri to some shared file so both usecases could use it, my mistake not doing that but it is not a must at all.

**Review by djabarovgeorge — APPROVED** [review]
Looks great to me, I run it locally and it worked as expected! 
Awesome job!🚀, you probably had a lot of drawbacks in the flow.

**Pull request merged**